### PR TITLE
Victoria Sawchuk -- Carets

### PR DESCRIPTION
**Pull request by vsawchuk**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |    When determining whether or not my Block class should inherit behavior from my Reservation class, I was originally tempted to make Block a subclass of Reservation, but I realized that while they had the same method names, their behavior was mostly different, so I made them separate classes with a single mix-in.   |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  Mix-ins and how to break out methods into smaller bits of functionality to better attempt compliance with SRP.  |
| Describe a _nominal test_ that you wrote for this assignment. |  Any of my tests that test for the type of data returned by a method - I did this to be sure I was always returning the expected data structures/elements in case I made any poorly executed changes in the future.  |
| Describe an _edge case test_ that you wrote for this assignment. |  Testing that I prevent blocked rooms from being reserved by the general public. I blocked all the rooms off, and then indicated that any reservation without a block_id would raise an Exception.   |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |  I didn't do a great job sticking to that - I much more prefer to write the actual code once I have a design in mind.  So it was hard for me to commit to writing pseudocode for everything.  |

### CODE REVIEW

File: lib/Custom_errors.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class NoRoomsAvailableError < StandardError
```
Comment (CheezItMan) [inline lib/Custom_errors.rb:1]: I like the use of custom errors.

File: lib/interface.rb [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  end
+
+  class Block
+    def self.clear
+      @@blocks = []
+    end
+  end
+end
+
+def valid_date?(input_date)
```
Comment (CheezItMan) [inline lib/interface.rb:25]: Very nice work making a regex to validate a date.

File: lib/Room.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    def collect_instance
+      @@rooms.push(self)
+    end
+
+    def self.all
+      return @@rooms
+    end
+
+    def self.update_price(room_num, new_rate)
```
Comment (CheezItMan) [inline lib/Room.rb]: Why not provide a way in the `initialize` method to start off a room with an optional parameter for price.

File: lib/Reservation.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      room_num = Reservation.sample_available_rooms(start_date, end_date, block_id).sample if room_num == 0
+      prevent_booking_blocked(room_num, block_id)
+      prevent_doublebooking(start_date, end_date, room_num)
+      @room_num = room_num
+      collect_instance
+    end
+
+    def self.all(date = nil)
+      return @@reservations if date == nil
+      return @@reservations.select { |reservation| reservation.start_date <= date && reservation.end_date >= date }
```
Comment (CheezItMan) [inline lib/Reservation.rb:23]: I really like the way you have this `.all` method written.  The `select` is a really nice touch.

File: specs/Reservation_spec.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    Hotel::Room.new(21)
+    Hotel::Reservation.new(nil, Date.today, Date.today + 1, 21).must_be_instance_of Hotel::Reservation
+  end
+
+  it "can calculate the total cost of the reservation" do
+    reservation = Hotel::Reservation.new(nil, Date.today, Date.today + 4)
+    reservation.total.must_equal 800
+  end
+
+  it "can provide information about reservations on a specific date" do
```
Comment (CheezItMan) [inline specs/Reservation_spec.rb:33]: This `it` block looks like it could be broken up into separate it blocks.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Lots of good commits, and the commit messages look pretty good  |
Answer comprehension questions	|  Check, it' understandable that you didn't stick to the pseudocode.  Plans change and shouldn't be too fixed to the earlier pseudocode/design.    |
**Design** |
Each class is responsible for a single piece of the program |  Well defined, I like the custom errors.  Interesting choice to use mix-ins for things like overlapping dates.  Why not instead do a `DateRange` class that   The classes are set for a single responsibility which is great!   | 
Classes are loosely coupled | They interact with each other through standardized `.all` methods and   |
**Wave 1** |
List rooms | You can list rooms and you can get a list of rooms available in a given date range. |
Reserve a room for a given date range |  Check by creating a Reservation object.  |
List reservations for a given date | Check, and tested in your reservation spec.  | 
Calculate reservation price |  Check, nice use of enumerable.   | 
Invalid date range produces an error | Check, good use of custom errors  |
**Wave 2** |
View available rooms for a given date range | Check |  
Reserving a room that is not available produces an error | Check  |
**Wave 3** |
Create a block of rooms | Check, nice use of sample. |
Check if a block has rooms |  Check  |
Reserve a room from a block | Check, and tested for the nominal case.  You covered the edge cases where you add a room to two different blocks, and where you try to reserve the same room in the same time twice.      |
Test coverage | 100% coverage! |
**Additional Feedback** | Very nice work.  You hit all the requirements.   |

**Review by CheezItMan — COMMENTED** [review]
Nice work, just a few comments in-line.
